Mathlib/Analysis/Polynomial/MahlerMeasure.lean

```diff
@@ -0,0 +1,126 @@
+`(2 * π)⁻¹ * ∫ x ∈ (0, 2 * π), log ‖p (e ^ (i * x))‖`.
+- `Polynomial.mahlerMeasure p`: the (exponential) Mahler measure of a polynomial `p`, which is equal
+to `e ^ p.logMahlerMeasure` if `p` is nonzero, and `0` otherwise.
+
+## Main results
+
+- `Polynomial.mahlerMeasure_mul`: the Mahler measure of the product of two polynomials is the
+product of their Mahler measures.
+-/
+
+namespace Polynomial
+
+open Real
+
+/-- The logarithmic Mahler measure of a polynomial `p` defined as
+`(2 * π)⁻¹ * ∫ x ∈ (0, 2 * π), log ‖p (e ^ (i * x))‖` -/
+noncomputable def logMahlerMeasure (p : ℂ[X]) := circleAverage (fun x ↦ log ‖eval x p‖) 0 1
+
+theorem logMahlerMeasure_def (p : ℂ[X]) : p.logMahlerMeasure =
+    circleAverage (fun x ↦ log ‖eval x p‖) 0 1 := rfl
+
+@[simp]
+theorem logMahlerMeasure_zero : (0 : ℂ[X]).logMahlerMeasure = 0 := by
+  simp [logMahlerMeasure_def, circleAverage_def]
+
+@[simp]
+theorem logMahlerMeasure_one : (1 : ℂ[X]).logMahlerMeasure = 0 := by
+  simp [logMahlerMeasure_def, circleAverage_def]
+
+@[simp]
+theorem logMahlerMeasure_const (z : ℂ) : (C z).logMahlerMeasure = log ‖z‖ := by
+  simp [logMahlerMeasure_def, circleAverage_def, mul_assoc]
+
+@[simp]
+theorem logMahlerMeasure_X : (X : ℂ[X]).logMahlerMeasure = 0 := by
+  simp [logMahlerMeasure_def, circleAverage_def]
+
+@[simp]
+theorem logMahlerMeasure_monomial (n : ℕ) (z : ℂ) : (monomial n z).logMahlerMeasure = log ‖z‖ := by
+  simp [logMahlerMeasure_def, circleAverage_def, mul_assoc]
+
+/-- The Mahler measure of a polynomial `p` defined as `e ^ (logMahlerMeasure p)` if `p` is nonzero
+and `0` otherwise -/
+noncomputable def mahlerMeasure (p : ℂ[X]) := if p ≠ 0 then exp (p.logMahlerMeasure) else 0
+
+theorem mahlerMeasure_def_of_ne_zero {p : ℂ[X]} (hp : p ≠ 0) : p.mahlerMeasure =
+    exp ((2 * π)⁻¹ * ∫ (x : ℝ) in (0)..(2 * π), log ‖eval (circleMap 0 1 x) p‖) :=
+  by simp [mahlerMeasure, hp, logMahlerMeasure_def, circleAverage_def]
+
+theorem logMahlerMeasure_eq_log_MahlerMeasure {p : ℂ[X]} :
+    p.logMahlerMeasure = log p.mahlerMeasure := by
+  rw [mahlerMeasure]
+  split_ifs <;> simp_all [logMahlerMeasure_def, circleAverage_def]
+
+@[simp]
+theorem mahlerMeasure_zero : (0 : ℂ[X]).mahlerMeasure = 0 := by simp [mahlerMeasure]
+
+@[simp]
+theorem mahlerMeasure_one : (1 : ℂ[X]).mahlerMeasure = 1 := by simp [mahlerMeasure]
+
+@[simp]
+theorem mahlerMeasure_const (z : ℂ) : (C z).mahlerMeasure = ‖z‖ := by
+  simp only [mahlerMeasure, ne_eq, map_eq_zero, logMahlerMeasure_const, ite_not]
+  split_ifs with h
+  · simp [h]
+  · simp [h, exp_log]
+
+theorem mahlerMeasure_nonneg (p : ℂ[X]) : 0 ≤ p.mahlerMeasure := by
+  by_cases hp : p = 0 <;> simp [hp, mahlerMeasure_def_of_ne_zero, exp_nonneg]
+
+@[simp]
+theorem mahlerMeasure_eq_zero_iff (p : ℂ[X]) : p.mahlerMeasure = 0 ↔ p = 0 := by
+  refine ⟨?_, by simp_all [mahlerMeasure_zero]⟩
+  contrapose
+  exact fun h ↦ by simp [mahlerMeasure_def_of_ne_zero h]
+
+private lemma mahlerMeasure_integrable (p : ℂ[X]) :
```

Comment:
```suggestion
lemma intervalIntegrable_mahlerMeasure (p : ℂ[X]) :
```
to use the same names as in the lemma statement. Also, I'd drop the `private` because I think this lemma could be used a lot when talking about Mahler measures.